lib/pdn/pdn.c

```diff
@@ -0,0 +1,763 @@
+	CMD_PDN_DEACTIVATE,
+	CMD_APN_DISCONNECT,
+	CMD_MAX
+};
+
+/** @brief AT command responses
+ * Internal definitions for supported AT command responses
+ */
+enum pdn_management_at_evt {
+	EVT_OK = 0,
+	EVT_CONTEXT_CREATE,
+	EVT_PDN_ID_GET,
+	EVT_PDN_STATE_GET,
+	EVT_CONTEXT_QUERY,
+	EVT_MAX
+};
+
+/** Data type for PDN connection. */
+struct pdn_connection {
+	/* Client defined callback when PDN related nofication is received */
+	pdn_notif_handler_t callback;
+	/* Modem assigned identifier for uniquely identifying a PDN. */
+	int context_id;
+	/* PDN id. */
+	unsigned int pdn_id;
+	/* Indicates the connection state #pdn_conn_state of the PDN. */
+	volatile unsigned int state;
+	/* APN Name. */
+	char p_apn[PDN_MAX_APN_NAME];
+	/* Supported address families.
+	 * Bit mask indicating PDN_AF_INET, PDN_AF_INET6 or both.
+	 */
+	sa_family_t family;
+};
+
+static const char *const m_pdn_mananagement_at_cmds[CMD_MAX] = {
+	[CMD_CONTEXT_CREATE] = "AT%%XNEWCID?",
+	[CMD_APN_CONNECT] = "AT+CGDCONT=",
+	[CMD_PDN_ACTIVATE] = "AT+CGACT=1,",
+	[CMD_PDN_ID_GET] = "AT%%XGETPDNID=",
+	[CMD_PDN_STATE_GET] = "AT+CGACT?",
+	[CMD_PDN_DEACTIVATE] = "AT+CGACT=0,",
+	[CMD_APN_DISCONNECT] = "AT+CGDCONT="
+};
+
+typedef int (*pdn_event_parser_t)(struct pdn_connection *const handle,
+				  char *const p_at_event);
+
+static int ok_rsp_parser(struct pdn_connection *const handle,
+			 char *const p_at_event);
+static int context_create_parser(struct pdn_connection *const handle,
+				 char *const p_at_event);
+static int pdn_id_get_parser(struct pdn_connection *const handle,
+			     char *const p_at_event);
+static int pdn_state_get_parser(struct pdn_connection *const handle,
+				char *const p_at_event);
+static int context_query_parser(struct pdn_connection *const handle,
+				char *const p_at_event);
+
+struct pdn_at_evt_t {
+	const char *p_at_str;
+	pdn_event_parser_t parser_fn;
+};
+
+static const struct pdn_at_evt_t m_pdn_mananagement_at_evts[EVT_MAX] = {
+	[EVT_OK] = { .p_at_str = "OK", .parser_fn = ok_rsp_parser },
+	[EVT_CONTEXT_CREATE] = { .p_at_str = "%XNEWCID:",
+				 .parser_fn = context_create_parser },
+	[EVT_PDN_ID_GET] = { .p_at_str = "%XGETPDNID:",
+			     .parser_fn = pdn_id_get_parser },
+	[EVT_PDN_STATE_GET] = { .p_at_str = "+CGACT:",
+				.parser_fn = pdn_state_get_parser },
+	[EVT_CONTEXT_QUERY] = { .p_at_str = "+CGDCONT:",
+				.parser_fn = context_query_parser },
+};
+
+static void _at_notif_handler(void *context, const char *response)
+{
+	if ((context != NULL) && (response != NULL)) {
+		struct pdn_connection *const p_pdn =
+			(struct pdn_connection *)context;
+		enum pdn_event event = PDN_EVENT_NA;
+
+		/* TODO: Validate received AT event here ...*/
+
+		if ((event != PDN_EVENT_NA) && (p_pdn->callback != NULL)) {
+			p_pdn->callback(p_pdn, event);
+		}
+	}
+}
+
+static struct pdn_connection *connection_allocate(void)
+{
+	/* Reserve PDN handle */
+	struct pdn_connection *const p_pdn =
+		malloc(sizeof(struct pdn_connection));
```

Comment:
use `k_malloc` and `k_free` instead.

Reply:
k_calloc in this case to avoid the the memset.